base/shell/explorer/syspager.cpp

```diff
@@ -1552,6 +1596,164 @@ LRESULT CSysPagerWnd::OnCopyData(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL&
     return FALSE;
 }
 
+BOOL CSysPagerWnd::EnumIcons(BOOL (*Operation)(IconWatcherData*, CSysPagerWnd *obj, BOOL Exp), LPARAM lParam, BOOL Expand)
+{
+    if (Operation == NULL || lParam == 0)
+        return FALSE;
+    
+    CAtlList<IconWatcherData *> *IconList = reinterpret_cast<CAtlList<IconWatcherData *> *>(lParam);
+
+    IconWatcherData *Entry = NULL;
+    POSITION NextPosition;
+
+    for (NextPosition = IconList->GetHeadPosition(); NextPosition != NULL;)
+    {
+        Entry = IconList->GetNext(NextPosition);
+        if (Entry && !Operation(Entry, this, Expand))
+        {
+            return TRUE;
+        }
+    }
+
+    return TRUE;
+}
+
+LRESULT CSysPagerWnd::IsActive(IconWatcherData *pIcon)
+{
+    /**
+     For now, simulate Volume and Remove hardware inactive
+    */
+    if (bAutoTrayEnabled == FALSE)
+        return TRUE;
+    else if (pIcon->IconData.uID == 33996 || pIcon->IconData.uID == 33997)
+        return FALSE;
+    return TRUE;
+}
+
+BOOL EnumProc(IconWatcherData *pIcon, CSysPagerWnd *obj, BOOL Expand)
+{
+    TBBUTTONINFO tbbi = { 0 };
+
+    tbbi.cbSize = sizeof(tbbi);
+    tbbi.dwMask = TBIF_STATE;
+    tbbi.fsState = TBSTATE_ENABLED;
+
+    int index = obj->Toolbar.FindItem(pIcon->IconData.hWnd, pIcon->IconData.uID, NULL);
+
+    if (!obj->IsActive(pIcon))
+    {
+        if (Expand)
+        {
+            if ((pIcon->IconData.dwState & NIS_HIDDEN) == NIS_HIDDEN)
+            {
+                pIcon->IconData.dwState &= ~NIS_HIDDEN;
+                tbbi.fsState &= ~TBSTATE_HIDDEN;
+                obj->Toolbar.IncVisibleButtons();
+            }
+        }
+        else if (pIcon->dwPreference != ALWAYS_SHOW )
+        {
+            if ((pIcon->IconData.dwState & NIS_HIDDEN) != NIS_HIDDEN)
+            {
+                pIcon->IconData.dwState |= NIS_HIDDEN;
+                tbbi.fsState |= TBSTATE_HIDDEN;
+                obj->Toolbar.DecVisibleButtons();
+            }
+        }
+    }
+    else if (pIcon->dwPreference != ALWAYS_HIDE)
+    {
+        if ((pIcon->IconData.dwState & NIS_HIDDEN) == NIS_HIDDEN)
+         {
+             pIcon->IconData.dwState &= ~NIS_HIDDEN;
+             tbbi.fsState &= ~TBSTATE_HIDDEN;
+             obj->Toolbar.IncVisibleButtons();
+         }
+    }
+    obj->Toolbar.SetButtonInfo(index, &tbbi);
+
+    return TRUE;
+}
+
+LRESULT CSysPagerWnd::OnMouseLeave(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    if (bExpanded)
+    {
+        AutoTray(FALSE);
+        bExpanded = FALSE;
+    }
+    return 0;
+}
+
+VOID CSysPagerWnd::AutoTray(BOOL expand)
+{
+
+    PullServicesEnabled();
+    EnumIcons(EnumProc, reinterpret_cast<LPARAM>(GetIconList()), expand);
```

Comment:
You are not iterating the actual active icons list,
but the list kept by an object that checks icons to see if one died.